orttraining/orttraining/python/training/ortmodule/_custom_autograd_function_runner.py

```diff
@@ -80,7 +76,7 @@ def generate_non_leaf_or_not(grad_flag, tensor_flag, arg, is_training_mode, is_i
         else:
             return arg
 
-    def wrap_all_outputs(result, training_mode_flag):
+    def wrap_all_outputs(result, tensors_generated_by_ort: List[torch.Tensor], training_mode_flag: bool):
```

Comment:
No annotation for the first one?

Reply:
BTW I saw your other comments, I was considering to run the forward pass once and collect all ortvalues that are saved in ctx, then for subsquent forward passes, we know exactly what tensors we need excpkictly copy before giving it to torch.autograd.Function.forward(). Doing this way, users code don't need to be aware of the copy but still be safe since we(ortmodule) copied for them.

Reply:
I was wondering if there was a way to avoid a copy. dlapck only implies ownership transfer so it cannot be used for every saved tensor. Maybe we could save the OrtValue instead of the torch tensor but it seemed a risky change.

Reply:
I am afraid, for those tensors that needs to be saved in context, we have to copy it, unless we let ORT allocation planner know there is such a reference, and know when it releases it. While knowing it requires ORT to run the forward pass once, which is behind the allocation plan generation (during session creation). ORT not ware of such a reference, then it will call torch allocator's API to release the memory, then the data got changed on that address. When backward runs, though ctx get unpack a tensor from the saved address, but the content is not expected any more.

Reply:
I understood that's what happen when ORT receives an OrtValue through the dlpack protocol. But what it receives an OrtValue already allocated (like run_with_ortvalues). torch des not need this copy. My guess was there should be a way to avoid it too when torch is calling ort. But the time may be not significant enough to think about that.

Reply:
FYI I updated the code, the details is in PR description.